src/arena.c

```diff
@@ -177,52 +364,36 @@ arena_large_dalloc_stats_update(arena_t *arena, size_t usize) {
 	index = size2index(usize);
 	hindex = (index >= NBINS) ? index - NBINS : 0;
 
-	arena->stats.ndalloc_large++;
-	arena->stats.allocated_large -= usize;
-	arena->stats.lstats[hindex].ndalloc++;
-	arena->stats.lstats[hindex].curlextents--;
+	arena_stats_add_u64(tsdn, &arena->stats, &arena->stats.ndalloc_large,
+	    1);
+	arena_stats_sub_zu(tsdn, &arena->stats, &arena->stats.allocated_large,
+	    usize);
+	arena_stats_add_u64(tsdn, &arena->stats,
+	    &arena->stats.lstats[hindex].ndalloc, 1);
+	arena_stats_sub_zu(tsdn, &arena->stats,
+	    &arena->stats.lstats[hindex].curlextents, 1);
 }
 
 static void
-arena_large_reset_stats_cancel(arena_t *arena, size_t usize) {
+arena_large_reset_stats_cancel(tsdn_t *tsdn, arena_t *arena, size_t usize) {
 	szind_t index = size2index(usize);
 	szind_t hindex = (index >= NBINS) ? index - NBINS : 0;
 
 	cassert(config_stats);
 
-	arena->stats.ndalloc_large--;
-	arena->stats.lstats[hindex].ndalloc--;
+	arena_stats_lock(tsdn, &arena->stats);
+	arena_stats_sub_u64(tsdn, &arena->stats, &arena->stats.ndalloc_large,
```

Comment:
It's unclear to me that why this section isn't cancelling the same changes as in arena_large_dalloc_stats_update().